### DISCUSSION TITLE
Checkout with private submodules doesn't work

### DISCUSSION THREAD

**Discussion by ShabalinAnton** [community/community#26936] [Actions]
<p>Hello. I can’t get to work private submodule checkout. </p>
<p>I’ve created two private repositories: first_repo and second_repo. </p>
<p>Then I’ve added second_repo as a submodule to first_repo.</p>
<p>When I am trying to make checkout it gives me the following:</p>
<pre><code>Setting up auth
"C:\Program Files\Git\bin\git.exe" config --local --name-only --get-regexp core\.sshCommand
"C:\Program Files\Git\bin\git.exe" submodule foreach --recursive "git config --local --name-only --get-regexp 'core\.sshCommand' &amp;&amp; git config --local --unset-all 'core.sshCommand' || :"
"C:\Program Files\Git\bin\git.exe" config --local --name-only --get-regexp http\.https\:\/\/github\.com\/\.extraheader
"C:\Program Files\Git\bin\git.exe" submodule foreach --recursive "git config --local --name-only --get-regexp 'http\.https\:\/\/github\.com\/\.extraheader' &amp;&amp; git config --local --unset-all 'http.https://github.com/.extraheader' || :"
"C:\Program Files\Git\bin\git.exe" config --local http.https://github.com/.extraheader "AUTHORIZATION: basic ***"

Fetching the repository
"C:\Program Files\Git\bin\git.exe" -c protocol.version=2 fetch --no-tags --prune --progress --no-recurse-submodules --depth=1 origin +refs/heads/master:refs/remotes/origin/master
Logon failed, use ctrl+c to cancel basic credential prompt.
&lt;font color="#FF0000"&gt;##[error]fatal: could not read Username for 'https://github.com': terminal prompts disabled &lt;/font&gt;The process 'C:\Program Files\Git\bin\git.exe' failed with exit code 128
Waiting 14 seconds before trying again
</code></pre>
<p>My yml file:</p>
<pre><code>- name: Checkout
      uses: actions/checkout@v2

    - name: Checkout private tools
      uses: actions/checkout@v2
      with:
        repository: ShabalinAnton/second_repo
        token: ${{ secrets.MY_SECRET_PAT }}
        path: second_repo
</code></pre>
<p>I have already tried almost every solution I’ve found. But nothing seems to be working. </p>
<p>Does anyone know the soluttion ?</p>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/shabalinanton">@shabalinanton</a> ,</p>
<p>I can repro the same on my side:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/9/916770d0cc9938bb58d4774d485b995e6475f084.png" data-download-href="/uploads/short-url/kKiQWfpPKdSMMfxu4Fg4FtF7ai8.png?dl=1" title="submodule2.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/52772761/181100873-030bf98d-2bd2-42f2-8ac2-2582bebc086b.png" alt="submodule2.png" data-base62-sha1="kKiQWfpPKdSMMfxu4Fg4FtF7ai8" width="690" height="242" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/9/916770d0cc9938bb58d4774d485b995e6475f084_2_690x242.png, https://aws1.discourse-cdn.com/github/optimized/2X/9/916770d0cc9938bb58d4774d485b995e6475f084_2_1035x363.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/9/916770d0cc9938bb58d4774d485b995e6475f084.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/9/916770d0cc9938bb58d4774d485b995e6475f084_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">submodule2.png</span><span class="informations">1298×456 34.2 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Please re-create the personal access token and update the value of secrets, confirmed it works:</p>
<p><img src="https://user-images.githubusercontent.com/52772761/181100874-79abd4f8-4322-43c9-8950-fe34ef6bf0f4.png" alt="submodule3.png" data-base62-sha1="zbcytsMiqupdK6VZmS6ll2mpCb0" width="534" height="158"></p>
<p>Actually you can directly checkout both repo with below code:</p>
<pre><code>- name: Checkout
        uses: actions/checkout@v2
        with:
           token: ${{ secrets.PAT1 }}
&lt;font color="#FF6600"&gt;submodules: true&lt;/font&gt;
</code></pre>
<p>Or you can clone the private repo via command:</p>
<pre><code>- name: checkout
         run: |
           git clone https://username:$PAT@github.com/{org}/{repo}.git
         env:
           PAT: ${{ secrets.PAT1 }}
</code></pre>
<p>Thanks.</p>

**Comment by ShabalinAnton** [comment]
<p>Hello, <a class="mention" href="/weide-zhou">@weide-zhou</a> Thanks for the answer!</p>
<p>But I still have a problem.</p>
<p>When I am trying to do:</p>
<pre><code>- name: checkout2
      shell: bash
      run: |
         git clone https://ShabalinAnton:$PAT@github.com/ShabalinAnton/second_repo.git
      env:
        PAT: '${{secrets.MY_PAT}}'
</code></pre>
<p>Then I get the following error:</p>
<pre><code>Run git clone ***github.com/ShabalinAnton/second_repo.git
  git clone ***github.com/ShabalinAnton/second_repo.git
  shell: /bin/bash --noprofile --norc -e -o pipefail {0}
  env:
    PAT:
Cloning into 'second_repo'...
remote: Invalid username or password.
&lt;font color="#FF0000"&gt;fatal: Authentication failed for 'https://github.com/ShabalinAnton/second_repo.git/'&lt;/font&gt;##[error]Process completed with exit code 128.
</code></pre>
<p>But everything work when I change ${{secrets.MY_PAT}} to the key value, for example:</p>
<pre><code>- name: checkout2
      shell: bash
      run: |
         git clone https://ShabalinAnton:$PAT@github.com/ShabalinAnton/second_repo.git
      env:
        PAT: 'a91c5f7477bb6234e30833c330815ce0d26f831d'
</code></pre>
<p>I believe something could be wrong with a syntax. Also this error appears when I am trying to use repository secret. </p>
<p>This is my yml file:</p>
<pre><code>name: Build And Deploy
on:
  push:
    branches: [master]

jobs:
  build:
    runs-on: windows-latest

    steps:
     - uses: actions/checkout@v2

     - uses: actions/checkout@v2
        with:
            submodules: 'true'
            token: '${{ secrets.SECOND_REPO_SECRET }}'
</code></pre>
<p>Which gives me the following result:</p>
<pre><code>/usr/bin/git -c protocol.version=2 fetch --no-tags --prune --progress --no-recurse-submodules --depth=1 origin +84a8987c1d0ede9770868a27ef4547c623df3f45:refs/remotes/origin/master
&lt;font color="#FF0000"&gt;##[error]fatal: could not read Username for 'https://github.com': terminal prompts disabled&lt;/font&gt; The process '/usr/bin/git' failed with exit code 128
  Waiting 16 seconds before trying again
  /usr/bin/git -c protocol.version=2 fetch --no-tags --prune --progress --no-recurse-submodules --depth=1 origin +84a8987c1d0ede9770868a27ef4547c623df3f45:refs/remotes/origin/master
  ##[error]fatal: could not read Username for 'https://github.com': terminal prompts disabled
  The process '/usr/bin/git' failed with exit code 128
  Waiting 19 seconds before trying again
  /usr/bin/git -c protocol.version=2 fetch --no-tags --prune --progress --no-recurse-submodules --depth=1 origin +84a8987c1d0ede9770868a27ef4547c623df3f45:refs/remotes/origin/master
  ##[error]fatal: could not read Username for 'https://github.com': terminal prompts disabled
  ##[error]The process '/usr/bin/git' failed with exit code 128
</code></pre>

**Comment by weide-zhou** [comment]
<p>Hi <a class="mention" href="/shabalinanton">@shabalinanton</a> ,</p>
<p>Based on your error message, ${{secrets.MY_PAT}} doesn’t have the value of your token.</p>
<p>Please update the secrets MY_PAT with the new value, or re-create a new PAT token from github setting <a href="https://github.com/settings/tokens" rel="nofollow noopener">https://github.com/settings/tokens,</a> and then update the secrets with the new PAT.</p>
<p>Please make sure the secrets <strong>exist on the repo which contains the workflow yaml</strong>.</p>
<p>Regarding the submodule parameter, you don’t need to checkout twice, only one action is enough, it will automatically checkout the submodule repo. Please check the usage <a href="https://github.com/actions/checkout#usage" rel="nofollow noopener">link</a> for more details.</p>
<pre><code> - uses: actions/checkout@v2 - uses: actions/checkout@v2
        with:
repository: {org}/{REPO_ONE}
            submodules: 'true'
            token: '${{ secrets.MY_PAT }}'
</code></pre>
<p>Thanks.</p>

**Comment by ShabalinAnton** [comment]
<p>Hello again, <a class="mention" href="/weide-zhou">@weide-zhou</a> </p>
<p>Please, watch the following. I’ve created a video with steps to reproduce my problem:</p>
<p><a href="https://youtu.be/EIJjueSbg2U" rel="nofollow noopener">https://youtu.be/EIJjueSbg2U</a> </p>
<p>I can’t get why it don’t work on my side. Maybe you can help me.</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/shabalinanton">@shabalinanton</a> ,</p>
<p>You didn’t store the PAT as the secret value of repo_one. Until the last minute of your video, you create the secrets, but with incorrect value…</p>
<p>Please update secert MY_SECRET with your PAT token.</p>
<p><img src="https://user-images.githubusercontent.com/52772761/181100877-71b5ca61-57ed-496b-a0b2-a00410c74664.png" alt="secretsvalue.png" data-base62-sha1="j5IFNevZmk3xNZqyRWgQxnWuHYq" width="676" height="53"></p>
<p>And in workflow yaml of repo_one:</p>
<pre><code>- name: Checkout private tools
         uses: actions/checkout@v2
         with:
           repository: ShabakinAnton/&lt;font color="#FF0000"&gt;first_repo&lt;/font&gt; token: ${{ secrets.MY_SECRET }}
           submodules: true
</code></pre>
<p>Please refer to official <a href="https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets" rel="nofollow noopener">doc</a> for more details.</p>
<p>Thanks</p>
